### ISSUE TITLE
Should we mask the attention before applying the softmax?

### ISSUE THREAD

**Issue by eduOS** [abisee/pointer-generator#88]
In [this line](https://github.com/abisee/pointer-generator/blob/master/attention_decoder.py#L99): 
`          attn_dist *= enc_padding_mask # apply mask`

of this function: 

```
def masked_attention(e):
    """Take softmax of e then apply enc_padding_mask and re-normalize"""
    attn_dist = nn_ops.softmax(e) # take softmax. shape (batch_size, attn_length)
    attn_dist *= enc_padding_mask # apply mask
    masked_sums = tf.reduce_sum(attn_dist, axis=1) # shape (batch_size)
    return attn_dist / tf.reshape(masked_sums, [-1, 1]) # re-normalize
```
you apply the mask after doing the softmax, then the probabilities would be summed up to less than 1. In my humble opinion, you it may be better to first mask the e and then do the softmax.

Does this make sense?

Best, 
Lerner

**Comment by eduOS** [comment]
It turns out that I am wrong.

**Comment by eduOS** [comment]
If you check the result of e^0, you'll know the answer.

**Comment by iamxpy** [comment]
@eduOS I am wondering on the same question now. I have read the implementation of Bert, where they DO *mask* the energy(scores) before applying the softmax, actually they fill the places where corresponding to PAD with a large negative number ,say -100000, which lead to softmax result (e^(-10000)) on those places will be all `0`. Personally I think this implementation makes more sense. Do you have any idea?
